### DISCUSSION TITLE
Did "yarp terminate" change its behaviour ?

### DISCUSSION THREAD

**Discussion by EliseiFrederic** [robotology/community#271] [Q&A] [answered]
Hi,

With our previous yarp version (2.3.64.8), we used yarp terminate a lot from the command line. For example to perform
yarp terminate /yarphear 
and that worked gracefully, whatever the targeted port.
So did it with our programs that implemented interruptModule() from RFMmodule.
It seems (our new version is 2.3.70.1) that it doesn't work anymore... Did the policy change ? 

Thanks in advance for the help.

**Comment by pattacini** [comment]
@Nicogene do you mind checking this?

**Comment by Nicogene** [comment]
Hi!
I didn't know about the existence of this command, I never used it but I just tried to run: 
```
$ yarpdev --device test_grabber
```
and 
```
$ yarp terminate /grabber 
```
it closed the `test_grabber`.
So I took a look to the code of `yarphear` and I find out that internally it is actually a `device` like the `test_grabber`. The devices in `YARP` have the "quit" port, you can connect to it to close the device.

**CONCLUSION**
`yarp terminate` connect to the "quit" port of the device and tell him to shut down, `yarphear` is a device (in particular `portaudio`) so `yarp terminate` works smoothly with it.
With a generic `RFModule` I'm afraid that it doesn't work, but I think it never worked.

If you have further doubts or questions don't hesitate to ask!

**Comment by EliseiFrederic** [comment]
Thanks for checking 👍  What you observe with your yarpdev is what I expect too...
That's what I get on my previous yarp install. Or if I have the yarp server running on my previous install.
But that's not what I get with my fresh install (updated to 2.3.70.2) !  

With the new version:
```sh 
$ yarpdev --device test_grabber
```
creates two visible ports on my yarp server: `/grabber` and `/test_grabber/quit`
```sh
$ yarp terminate /grabber does nothing
$ yarp terminate /test_grabber does nothing
$ yarp terminate /test_grabber/quit does the expected thing though! 
```
This is not only a device problem... On the old version, I can do
```sh
$ yarp write /test
```
and get it closing nicely with a `yarp terminate /test`
(there exists no `/test/quit` for this example...)

**Comment by Nicogene** [comment]
I tried it with yarp `2.3.70.2` and `2.3.71` and in both cases it works for me, with `yarpdev` and `yarp write` either. That's quite strange 🤔 

BTW let's analyse how `yarp terminate` works.

I took a deeper look to the command `terminate` and in `libYARP_OS/Companion.cpp` there is the function `cmdTerminate` that calls `Terminator::terminateByName(argv[0]);`.
In `terminateByName` there is this check:

``` c++
    if (s.find("/quit")==ConstString::npos) {
        // name doesn't include /quit
        // old mechanism won't work, let's try new
        PortCommand pc('\0', "i");
        Companion::sendMessage(s, pc, true);
        return true;
    }
```
That means that if you **don't** call `yarp terminate` on a "quit" port, it uses the "new" approach.
The `PortCommand pc('\0', "i");` is sent to the receiver and "i" is interpreted as follow:

(PortCoreInputUnit.cpp)
```c++
        case 'i':
            printf("Interrupt requested\n");
            //yarp::os::impl::kill(0, 2); // SIGINT
            //yarp::os::impl::kill(Logger::get().getPid(), 2); // SIGINT
            yarp::os::impl::kill(Logger::get().getPid(), 15); // SIGTERM
            break;
``` 

Do you compile from sources or do you use binaries? Which OS do you use? Do you compile yarp with `ACE` or without `ACE`?

**Comment by EliseiFrederic** [comment]
Thanks for looking to these details. We are using Linux (Jessie).
Our yarp is built from the sources.

If I send "TERM" signal to the "yarp write /test" process, it reacts as Interrupting, as expected from a "yarp terminate" command.

libACE-6.2.8 is available on the laptop, but I don't know if it's used to build yarp. Here are our flags : 
 ALLOW_IDL_GENERATION             OFF                                                                                                                    
 CMAKE_BUILD_TYPE                 Release                                                                                                                
 CMAKE_INSTALL_PREFIX             /usr/local                                                                                                             
 CREATE_DEVICE_LIBRARY_MODULES    OFF                                                                                                                    
 CREATE_GUIS                      ON                                                                                                                     
 CREATE_LIB_MATH                  ON                                                                                                                     
 CREATE_OPTIONAL_CARRIERS         ON                                                                                                                     
 CREATE_YARPBATTERYGUI            OFF                                                                                                                    
 CREATE_YARPDATADUMPER            ON                                                                                                                     
 CREATE_YARPDATAPLAYER            ON                                                                                                                     
 CREATE_YARPLASERSCANNERGUI       OFF                                                                                                                    
 CREATE_YARPLOGGER                ON                                                                                                                     
 CREATE_YARPMANAGER               ON                                                                                                                     
 CREATE_YARPMANAGER_CONSOLE       ON                                                                                                                     
 CREATE_YARPMOTORGUI              ON                                                                                                                     
 CREATE_YARPROBOTINTERFACE        ON                                                                                                                     
 CREATE_YARPSCOPE                 ON                                                                                                                     
 CREATE_YARPVIEW                  ON                                                                                                                     
 CUDA_BUILD_CUBIN                 OFF                                                                                                                    
 CUDA_BUILD_EMULATION             OFF                                                                                                                    
 CUDA_HOST_COMPILER               /usr/bin/cc                                                                                                            
 CUDA_SDK_ROOT_DIR                CUDA_SDK_ROOT_DIR-NOTFOUND                                                                                             
 CUDA_SEPARABLE_COMPILATION       OFF                                                                                                                    
 CUDA_TOOLKIT_ROOT_DIR            CUDA_TOOLKIT_ROOT_DIR-NOTFOUND                                                                                         
 CUDA_VERBOSE_BUILD               OFF                                                                                                                    
 DOXYGEN_PLANTUM_JAR              DOXYGEN_PLANTUM_JAR-NOTFOUND                                                                                           
 ENABLE_YARPRUN_LOG               OFF                                                                                                                    
 ENABLE_yarpcar_bayer             ON                                                                                                                     
 ENABLE_yarpcar_depthimage        OFF                                                                                                                    
 ENABLE_yarpcar_human             OFF                                                                                                                    
 ENABLE_yarpcar_mjpeg             OFF                                                                                                                    
 ENABLE_yarpcar_mpi               OFF                                                                                                                    
 ENABLE_yarpcar_mpibcast          OFF                                                                                                                    
 ENABLE_yarpcar_portmonitor       ON                                                                                                                     
 ENABLE_yarpcar_priority          ON                                                                                                                     
 ENABLE_yarpcar_rossrv            ON                                                                                                                     
 ENABLE_yarpcar_tcpros            ON                                                                                                                     
 ENABLE_yarpcar_xmlrpc            ON                                                                                                                     
 Eigen3_DIR                       Eigen3_DIR-NOTFOUND                                                                                                    
 FTDI_INCLUDE_DIR                 FTDI_INCLUDE_DIR-NOTFOUND                                                                                              
 FTDI_LIBRARY                     FTDI_LIBRARY-NOTFOUND                                                                                                  
 GLFW3_GLFW_LIBRARY               GLFW3_GLFW_LIBRARY-NOTFOUND                                                                                            
 GLFW3_INCLUDE_DIR                GLFW3_INCLUDE_DIR-NOTFOUND                                                                                             
 MPI_DEBUG_MSG                    OFF                                                                                                                    
 MPI_EXTRA_LIBRARY                /usr/lib/libmpi.so;/usr/lib/x86_64-linux-gnu/libdl.so;/usr/lib/x86_64-linux-gnu/libhwloc.so                            
 MPI_LIBRARY                      /usr/lib/libmpi_cxx.so                                                                                                 
 NITE2_INCLUDE_LOCAL              NITE2_INCLUDE_LOCAL-NOTFOUND                                                                                           
 NITE2_LIBRARY                    NITE2_LIBRARY-NOTFOUND                                                                                                 
 NVIDIACg_CgGL_LIBRARY            NVIDIACg_CgGL_LIBRARY-NOTFOUND                                                                                         
 NVIDIACg_GLEW_LIBRARY            NVIDIACg_GLEW_LIBRARY-NOTFOUND                                                                                         
 NVIDIACg_GL_LIBRARY              /usr/lib/x86_64-linux-gnu/libGL.so                                                                                     
 NVIDIACg_INCLUDE_DIR             NVIDIACg_INCLUDE_DIR-NOTFOUND                                                                                          
 NVIDIACg_glut_LIBRARY            /usr/lib/x86_64-linux-gnu/libglut.so                                                                                   
 NVIDIACg_pthread_LIBRARY         /usr/lib/x86_64-linux-gnu/libpthread.so                                                                                
 OpenCV_DIR                       /usr/share/OpenCV                                                                                                      
 OpenNI2_INCLUDE_DIR              OpenNI2_INCLUDE_DIR-NOTFOUND                                                                                           
 OpenNI2_LIBRARY                  OpenNI2_LIBRARY-NOTFOUND                                                                                               
 Qt5Core_DIR                      /usr/lib/x86_64-linux-gnu/cmake/Qt5Core                                                                                
 Qt5Gui_DIR                       /usr/lib/x86_64-linux-gnu/cmake/Qt5Gui                                                                                 
 Qt5Multimedia_DIR                /usr/lib/x86_64-linux-gnu/cmake/Qt5Multimedia                                                                          
 Qt5Network_DIR                   /usr/lib/x86_64-linux-gnu/cmake/Qt5Network                                                                             
 Qt5PrintSupport_DIR              /usr/lib/x86_64-linux-gnu/cmake/Qt5PrintSupport                                                                        
 Qt5Qml_DIR                       /usr/lib/x86_64-linux-gnu/cmake/Qt5Qml                                                                                 
 Qt5Quick_DIR                     /usr/lib/x86_64-linux-gnu/cmake/Qt5Quick                                                                               
 Qt5Widgets_DIR                   /usr/lib/x86_64-linux-gnu/cmake/Qt5Widgets                                                                             
 Qt5Xml_DIR                       /usr/lib/x86_64-linux-gnu/cmake/Qt5Xml                                                                                 
 Qt5_DIR                          /usr/lib/x86_64-linux-gnu/cmake/Qt5                                                                                    
 RTF_DIR                          RTF_DIR-NOTFOUND                                                                                                       
 SDLMAIN_LIBRARY                  /usr/lib/x86_64-linux-gnu/libSDLmain.a                                                                                 
 SDL_INCLUDE_DIR                  /usr/include/SDL                                                                                                       
 SDL_LIBRARY                      /usr/lib/x86_64-linux-gnu/libSDLmain.a;/usr/lib/x86_64-linux-gnu/libSDL.so;-lpthread                                   
 YARP_COMPILE_BINDINGS            OFF                                                                                                                    
 YCM_DIR                          YCM_DIR-NOTFOUND                                                                                                       
 wxWidgets_CONFIG_EXECUTABLE      wxWidgets_CONFIG_EXECUTABLE-NOTFOUND                                                                                   
 wxWidgets_wxrc_EXECUTABLE        wxWidgets_wxrc_EXECUTABLE-NOTFOUND

**Comment by drdanz** [comment, marked as answer]
The 'i' argument was working on linux only and it was considered harmful, therefore it was enabled only when YARP is built in Debug mode since yarp 2.3.70, see [here](https://github.com/robotology/yarp/blob/master/src/libYARP_OS/src/PortCoreInputUnit.cpp#L332-L339) and robotology/yarp@3f9ef5d1ae0861b15c56c75c9707c70281f779a9

As a consequence, the yarp terminate command now works only on the "/quit" ports, unless YARP is built in Debug mode.

**Comment by EliseiFrederic** [comment]
Thanks for pointing this change ! No more mystery now
